### PR TITLE
feat(synced-lyrics): multiple lyric sources

### PR DESCRIPTION
**Pull request by ArjixWasTaken**
Sneak Peek:

https://github.com/user-attachments/assets/29b317e8-3c1a-4a2f-ac9b-763f1c1c672c

🚧 **TODO** 🚧:
- [x] restructure the code to support multiple providers
- - [x] per provider lyrics
- - [x] per provider error messages
- - [x] per provider loading indicator
- - [x] per provider retry button
- [x] providers
- - [x] LRCLib
- - [x] Megalobiz
- - [x] Genius (??)
- - [ ] Musixmatch (??)
- - [x] YT Music (??)
- [ ] Customizable provider priority
- [ ] Semi-permanent lyric caching (localStorage??)
- [ ] Allow setting an offset to correct the timings of the lyrics (??)
- [ ] Allow enabling/disabling providers (looking at you `megalobiz`)

*may add more to this list at a later date*

### CODE REVIEW

**Comment by ArjixWasTaken** [comment]
This PR is not ready for any kind of review, currently no thought has been put into the code design since I was merely prototyping.
I will flesh this out in the coming days.

File: src/plugins/synced-lyrics/renderer/components/LyricsPicker.tsx [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+import type { YtIcons } from '@/types/icons';
+import { providers } from '@/plugins/synced-lyrics/providers';
+
+export const [providerIdx, setProviderIdx] = createSignal(0);
+export const LyricsPicker = () => {
+  const next = () => setProviderIdx((i) => (i + 1) % providers.length);
+  const previous = () => setProviderIdx((i) => (i - 1 + providers.length) % providers.length);
+
+  const chevronLeft: YtIcons = 'yt-icons:chevron_left';
+  const chevronRight: YtIcons = 'yt-icons:chevron_right';
```
Comment (ArjixWasTaken) [inline src/plugins/synced-lyrics/renderer/components/LyricsPicker.tsx:99]: as seen here, I have saved all the icon names as string unions, so it is easy to discover/use an icon

**Comment by ArjixWasTaken** [comment]
My exams are soon going to end (in 8 days), so I will be more active then.

**Comment by Rairof** [comment]
https://github.com/th-ch/youtube-music/pull/2383#issue-2499644176
I'd personally like those lyric sources either be kept hidden in the plugin settings or somewhere outside of the lyrics pace to not break the immersion rather then just being over the lyrics like that. Just my opinion.

**Comment by ArjixWasTaken** [comment]
The lyrics picker is not sticky, when you scroll down it will not be in view...

What is ur argument?

File: src/plugins/synced-lyrics/providers/index.ts [hunk 2]
```diff
@@ -20,27 +22,30 @@ interface SearchCache {
   >;
 }
 
+// prettier-ignore
+const initialData = () => Object.fromEntries(providers.map((p) => [p.name, { state: 'fetching' }]));
 const [accessor, setter] = createSignal<SearchCache>({
```
Comment (JellyBrick) [inline src/plugins/synced-lyrics/providers/index.ts]: I recommend `createStore` and `createResource` instead of `createSignal` for this situation

**Comment by Rairof** [comment]
https://github.com/th-ch/youtube-music/pull/2383#issuecomment-2394735575
I had no intentions of establishing any arguments. I just thought having a simple clean lyrics animation space like on Spotify is the way to go.
Example-
354297518-3d26b44a-2906-4036-afb1-a81d29960c9c

**Comment by ArjixWasTaken** [comment]
@Rairof 

What do you think of this?
I decided to make it sticky, and hide it when scrolling down.

https://github.com/user-attachments/assets/47600ab9-87ec-4680-92ac-31588845ce91

**Comment by Rairof** [comment]
https://github.com/th-ch/youtube-music/pull/2383#issuecomment-2424250600
Great! 
It's nice that now it's sticky. Would it be too much to make them pill shaped toggles similar to the "Song/video" toggles on left side?

**Comment by ArjixWasTaken** [comment]
> https://github.com/th-ch/youtube-music/pull/2383#issuecomment-2424250600
> Great! 
> It's nice that now it's sticky. Would it be too much trouble to make them pill shaped toggles similar to the "Song/video" toggles on left side that automatically fades away (hidden) after few secs and can be visible again upon hovering over the area? 

I am not sure I understand your suggestion, can you visualize it in any way?

**Comment by Rairof** [comment]
https://github.com/th-ch/youtube-music/pull/2383#issuecomment-2424485501
Sure.

**Comment by ArjixWasTaken** [comment]
Ooh, now I understand!
I am not really a fan of using pills like that.

Pills either mean a toggle between two states, like a switch, or a filter, like multiple-select.

Having pills used in the same fashion as tabs feels weird to me.

What are your other two suggestions?

**Comment by Rairof** [comment]
https://github.com/th-ch/youtube-music/pull/2383#issuecomment-2424596545
Alright no problem but can you make them hoverable (i.e it auto hides and the blank space is covered up until you hover over the are its supposed to be with your mouse then it pops up) instead of sticky as I think that would be better visually?
Suggestions-
1. It would be better to just have YT music source as default for lyrics, just makes more sense.
2. Auto switching sources each time lyrics aren't available for any sone without user having to manually retry or switch sources.

**Comment by ArjixWasTaken** [comment]
Ur 2nd suggestion is already part of the code

**Comment by ArjixWasTaken** [comment]
Having YouTube music as a lyric provider is possible, but to get synced lyrics I have to interface with the mobile API

Some paranoid people would fear getting banned by YouTube for that.

So I don't think it will be a default, heck, it may be an opt-in provider

**Comment by Rairof** [comment]
https://github.com/th-ch/youtube-music/pull/2383#issuecomment-2424623463
I see. Well that makes sense. That's all, thank you!

**Comment by ArjixWasTaken** [comment]
I like where this is going (code-wise), thanks for the store suggestion @JellyBrick !
Sadly I am dealing with a new bug, I suspect the automatic subscription to store changes is not working correctly.

Alas, I will try to resolve it, since the store does feel more intuitive and is less messy.

**Comment by ArjixWasTaken** [comment]
Progress!!

https://github.com/user-attachments/assets/4cc56653-053e-4c6c-844f-f7815f3b0c43

**Comment by ArjixWasTaken** [comment]
New provider `Megalobiz`!

image

**Comment by Rairof** [comment]
Suggestion- Can you set this section (in the pic below) to auto hide as in fade away or scroll up etc after a certain time (can be configured in settings)?
image

**Comment by ArjixWasTaken** [comment]
@Rairof as displayed in my previous videos, it hides away when you scroll down (auto-scroll also affects it)
can you explain why that ain't enough?

**Comment by Rairof** [comment]
https://github.com/th-ch/youtube-music/pull/2383#issuecomment-2435718702
Its not like it's not enough, I just think having a mouse hovering only approach to it instead of it being sticky is a better approach to a more uninterrupted visual experience.

**Comment by ArjixWasTaken** [comment]
Honestly I suck at CSS, I barely managed to get this far.
If anyone wants to give it a go I'd be thankful, I also wanted to have an appear-on-hover effect.

**Comment by janisslsm** [comment]
> Honestly I suck at CSS, I barely managed to get this far. If anyone wants to give it a go I'd be thankful, I also wanted to have an appear-on-hover effect.

```
.lyrics-renderer-sticky {
  position: sticky;
  top: var(--top, 0);
  z-index: 100;
  background-color: var(--ytmusic-background);

  transition: top 0.5s ease-in-out;
  opacity: 0;
  transition: opacity 1s;

  &:hover {
    opacity: 1;
  }
}
```

this achieves a fade-in/fade-out effect

https://github.com/user-attachments/assets/6f7e4f62-90da-457a-800a-daa3470f4fe0

**Comment by ArjixWasTaken** [comment]
Ok fine, you motivated me to fight with css and make what I envisioned.

https://github.com/user-attachments/assets/487a5800-cb80-414c-875d-b5e78599f98f

**Comment by Rairof** [comment]
can you make the lyrics picker react to mouse movement on top as well?

**Comment by ArjixWasTaken** [comment]
could you add a square outline on the areas you want to hover for the picker to be visible?

**Comment by Rairof** [comment]
https://github.com/th-ch/youtube-music/pull/2383#issuecomment-2436296110 also when hidden, the space could be filled up with lyrics
image

**Comment by ArjixWasTaken** [comment]
No I don't think I will.

**Comment by ArjixWasTaken** [comment]
image

**Comment by ArjixWasTaken** [comment]
Since a lot of progress has been done, I will open this up for review.
Just note that it ain't finished yet.

**Comment by ArjixWasTaken** [comment]
I've finally added LyricsGenius
image

question: since the lyrics are not synced line-by-line, should I make it auto-scroll based on the playback percentage?

**Comment by ArjixWasTaken** [comment]
I did some experiments, but automatically scrolling for plain lyrics doesn't work well.
I won't include it even as an option as it is really hit or miss.

Songs with minimal lyric-pauses sync well, but song's where the singer is not singing for a considerable amount of time are shitty with auto-scrolling.

I believe not having it at all is better.

**Comment by ArjixWasTaken** [comment]
Due to [`c1eead3` (#2383)](https://github.com/th-ch/youtube-music/pull/2383/commits/c1eead3607061a4755814f77bd67997c3f8b5168) I will be making this a draft again.

I am not interested in immediately fixing it, so I wanted to make sure this is not forgotten by mistake.

**Comment by ArjixWasTaken** [comment]
Latest preview:
https://github.com/user-attachments/assets/a3d6bbb3-d88c-404a-a4f8-93bd3f4b7187

**Comment by ArjixWasTaken** [comment]
I don't think I am going to implement musixmatch in this PR, so the current TODO list would be:

- [ ] Customizable provider priority
- [ ] Semi-permanent lyric caching (localStorage??)
- [ ] Allow setting an offset to correct the timings of the lyrics (??)
- [ ] Allow enabling/disabling providers (looking at you `megalobiz`)

Items with `(??)` are optional.

File: src/plugins/synced-lyrics/providers/YTMusic.ts [hunk 3]
```diff
@@ -0,0 +1,172 @@
+import { LyricProvider, LyricResult, SearchSongInfo } from '../types';
+
+const headers = {
+  Accept: 'application/json',
+  'Content-Type': 'application/json',
+};
+
+const client = {
```
Comment (JellyBrick) [inline src/plugins/synced-lyrics/providers/YTMusic.ts:8]: use the youtubei.js instead of this

**Review by JellyBrick — CHANGES_REQUESTED** [review]

File: src/plugins/synced-lyrics/providers/YTMusic.ts (hunk 3 above)
Comment (ArjixWasTaken) [reply src/plugins/synced-lyrics/providers/YTMusic.ts:8]: I did spend a considerable amount of time trying to do just that.

But it doesn't allow me to modify the client information.

Comment (JellyBrick) [reply src/plugins/synced-lyrics/providers/YTMusic.ts:8]: - If `youtubei.js` is impossible to use, it must have a way to use the user's YouTube session (using cookies).
  - In countries like South Korea, YTM features (including the `/next` endpoint) such as lyrics, song information, song playback, and search are unavailable without subscribing to YouTube Premium.
    - If the account does not have YouTube Premium, or if the session is not signed in, the response will always be empty

Comment (ArjixWasTaken) [reply src/plugins/synced-lyrics/providers/YTMusic.ts:8]: I will give it a try

**Comment by ArjixWasTaken** [comment]
I promise I didn't forget this PR, I just had no time lately.

**Comment by ArjixWasTaken** [comment]
@JellyBrick 

Before you make a new release, please give me until the end of this week to tidy up this PR.

**Comment by ArjixWasTaken** [comment]
I am working on this today, I was busy working on a uni assignment 😔

File: src/plugins/synced-lyrics/providers/YTMusic.ts (hunk 3 above)
Comment (ArjixWasTaken) [reply src/plugins/synced-lyrics/providers/YTMusic.ts:8]: I used UrbanVPN to connect to a Seoul server, and visited yt music from incognito and stuff like search and plain lyrics worked.
Does youtube not block those endpoints for VPNs?

Comment (ArjixWasTaken) [reply src/plugins/synced-lyrics/providers/YTMusic.ts:8]: Sadly I don't have an immediate solution, but I do have an idea in mind to resolve this issue forever™️

I am thinking of making a backend that fetches lyrics on demand and then caches them, to act like a proxy but just for lyrics.

Comment (JellyBrick) [reply src/plugins/synced-lyrics/providers/YTMusic.ts:8]: > I used UrbanVPN to connect to a Seoul server, and visited yt music from incognito and stuff like search and plain lyrics worked. Does youtube not block those endpoints for VPNs?

That IP band may have been recognized as a country other than South Korea.

Comment (JellyBrick) [reply src/plugins/synced-lyrics/providers/YTMusic.ts:8]: I will implement a cookie parser for youtubei when I have time.

Comment (ArjixWasTaken) [reply src/plugins/synced-lyrics/providers/YTMusic.ts:8]: The issue is not the cookie, it's having an authenticated session with the iOS client

Unless I've missed smth and you can reuse the web cookie for the iOS client.

**Review by JellyBrick — APPROVED** [review]

**Comment by Rairof** [comment]
@ArjixWasTaken 
Woops, might have missed the `Ambient mode` full screen theming compatibility?

image

**Comment by Rairof** [comment]
https://github.com/th-ch/youtube-music/pull/2383#issuecomment-2562873017
Works as of 3.7.1 thanks!
image

**Pull request merged**